### PR TITLE
Flatten fwd/fun and finish fwd flattening

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

This PR flattens fwd/fun and finishes flattening the fwd folder by creating the stan/math/fwd.hpp header. This again touches a lot of files, but the changes are very simple. Only the sum, log_sum_exp and to_fvar files required actual manual merging. The rest were just moves.

The new master header stan/math/fwd.hpp includes the following, which resembles what fwd/arr.hpp, fwd/mat.hpp, fwd/scal.hpp included (as well as the order): 

```
#include <stan/math/prim/mat/fun/Eigen.hpp>

#include <stan/math/fwd/core.hpp>
#include <stan/math/fwd/meta.hpp>
#include <stan/math/fwd/vectorize.hpp>
#include <stan/math/prim/arr.hpp>
#include <stan/math/prim/mat.hpp>
#include <stan/math/prim/scal.hpp>

#include <stan/math/fwd/fun.hpp>
#include <stan/math/fwd/functor.hpp>
```

## Tests

/
## Side Effects

/

## Checklist

- [x] Math issue #937 

- [x] Copyright holder: Rok Češnovar

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: stan/math/fwd/fun/mdivide_left.hpp [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
 
 #include <stan/math/prim/err.hpp>
 #include <stan/math/prim/mat/fun/Eigen.hpp>
 #include <stan/math/prim/mat/fun/typedefs.hpp>
 #include <stan/math/fwd/core.hpp>
-#include <stan/math/fwd/mat/fun/multiply.hpp>
+#include <stan/math/fwd/fun/multiply.hpp>
 #include <stan/math/prim/mat/fun/multiply.hpp>
-#include <stan/math/fwd/mat/fun/to_fvar.hpp>
+#include <stan/math/fwd/fun/to_fvar.hpp>
```
Comment (mcol) [inline stan/math/fwd/fun/mdivide_left.hpp]: Ordering.

File: stan/math/fwd/fun/mdivide_right.hpp [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
 #include <stan/math/prim/mat/fun/Eigen.hpp>
 #include <stan/math/prim/mat/fun/mdivide_right.hpp>
 #include <stan/math/fwd/core.hpp>
-#include <stan/math/fwd/mat/fun/typedefs.hpp>
-#include <stan/math/fwd/mat/fun/multiply.hpp>
+#include <stan/math/fwd/fun/typedefs.hpp>
+#include <stan/math/fwd/fun/multiply.hpp>
 #include <stan/math/prim/mat/fun/multiply.hpp>
-#include <stan/math/fwd/mat/fun/to_fvar.hpp>
+#include <stan/math/fwd/fun/to_fvar.hpp>
```
Comment (mcol) [inline stan/math/fwd/fun/mdivide_right.hpp]: Ordering.

File: stan/math/fwd/fun/multiply_lower_tri_self_transpose.hpp [hunk 3]
```diff
@@ -1,8 +1,8 @@
-#ifndef STAN_MATH_FWD_MAT_FUN_MULTIPLY_LOWER_TRI_SELF_TRANSPOSE_HPP
-#define STAN_MATH_FWD_MAT_FUN_MULTIPLY_LOWER_TRI_SELF_TRANSPOSE_HPP
+#ifndef STAN_MATH_FWD_FUN_MULTIPLY_LOWER_TRI_SELF_TRANSPOSE_HPP
+#define STAN_MATH_FWD_FUN_MULTIPLY_LOWER_TRI_SELF_TRANSPOSE_HPP
 
 #include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/fwd/mat/fun/multiply.hpp>
+#include <stan/math/fwd/fun/multiply.hpp>
 #include <stan/math/prim/mat/fun/transpose.hpp>
```
Comment (mcol) [inline stan/math/fwd/fun/multiply_lower_tri_self_transpose.hpp]: Ordering.

File: stan/math/fwd/fun/squared_distance.hpp [hunk 4]
```diff
@@ -1,9 +1,9 @@
-#ifndef STAN_MATH_FWD_MAT_FUN_SQUARED_DISTANCE_HPP
-#define STAN_MATH_FWD_MAT_FUN_SQUARED_DISTANCE_HPP
+#ifndef STAN_MATH_FWD_FUN_SQUARED_DISTANCE_HPP
+#define STAN_MATH_FWD_FUN_SQUARED_DISTANCE_HPP
 
 #include <stan/math/prim/err.hpp>
 #include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/fwd/mat/fun/dot_self.hpp>
+#include <stan/math/fwd/fun/dot_self.hpp>
 #include <stan/math/prim/mat/fun/subtract.hpp>
```
Comment (mcol) [inline stan/math/fwd/fun/squared_distance.hpp]: Ordering.

File: stan/math/fwd/fun/tcrossprod.hpp [hunk 5]
```diff
@@ -1,8 +1,8 @@
-#ifndef STAN_MATH_FWD_MAT_FUN_TCROSSPROD_HPP
-#define STAN_MATH_FWD_MAT_FUN_TCROSSPROD_HPP
+#ifndef STAN_MATH_FWD_FUN_TCROSSPROD_HPP
+#define STAN_MATH_FWD_FUN_TCROSSPROD_HPP
 
 #include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/fwd/mat/fun/multiply.hpp>
+#include <stan/math/fwd/fun/multiply.hpp>
 #include <stan/math/prim/mat/fun/transpose.hpp>
```
Comment (mcol) [inline stan/math/fwd/fun/tcrossprod.hpp]: Ordering.

File: stan/math/fwd/fun/to_fvar.hpp [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+template <typename T>
+inline fvar<T> to_fvar(const T& x) {
+  return fvar<T>(x);
+}
+
+/**
+ * Specialization of to_fvar for const fvars
+ *
+ *
```
Comment (mcol) [inline stan/math/fwd/fun/to_fvar.hpp]: A single empty line is enough (there are other like this in this file).

File: test/unit/math/fwd/fun/value_of_test.cpp [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+  vector<double> a_vals;
+  for (size_t i = 0; i < 10; ++i)
+    a_vals.push_back(i + 1);
+
+  vector<double> b_vals;
+  for (size_t i = 10; i < 15; ++i)
+    b_vals.push_back(i + 1);
+
+  vector<fvar<double> > a_fd;
```
Comment (mcol) [inline test/unit/math/fwd/fun/value_of_test.cpp]: [optional] Should we change `> >` to `>>` in this file?

File: stan/math/fwd.hpp [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+#include <stan/math/fwd/core.hpp>
+#include <stan/math/fwd/meta.hpp>
+#include <stan/math/fwd/vectorize.hpp>
+#include <stan/math/prim/arr.hpp>
+#include <stan/math/prim/mat.hpp>
+#include <stan/math/prim/scal.hpp>
+
+#include <stan/math/fwd/fun.hpp>
+#include <stan/math/fwd/functor.hpp>
```
Comment (mcol) [inline stan/math/fwd.hpp]: Wouldn't it make more sense to put the fwd headers together?

File: stan/math/fwd/fun/grad_inc_beta.hpp [hunk 9]
```diff
@@ -1,15 +1,15 @@
...
+#include <stan/math/fwd/fun/log.hpp>
+#include <stan/math/fwd/fun/log1m.hpp>
+#include <stan/math/fwd/fun/beta.hpp>
+#include <stan/math/fwd/fun/exp.hpp>
+#include <stan/math/fwd/fun/value_of.hpp>
 #include <stan/math/prim/scal/fun/value_of.hpp>
 #include <stan/math/fwd/core.hpp>
-#include <stan/math/fwd/scal/fun/inc_beta.hpp>
+#include <stan/math/fwd/fun/inc_beta.hpp>
 #include <stan/math/prim/scal/fun/grad_2F1.hpp>
```
Comment (mcol) [inline stan/math/fwd/fun/grad_inc_beta.hpp]: Ordering.

File: stan/math/fwd/fun/inc_beta.hpp [hunk 10]
```diff
@@ -1,12 +1,12 @@
...
 #include <stan/math/prim/scal/fun/grad_reg_inc_beta.hpp>
-#include <stan/math/fwd/scal/fun/pow.hpp>
-#include <stan/math/fwd/scal/fun/exp.hpp>
-#include <stan/math/fwd/scal/fun/beta.hpp>
-#include <stan/math/fwd/scal/fun/digamma.hpp>
+#include <stan/math/fwd/fun/pow.hpp>
+#include <stan/math/fwd/fun/exp.hpp>
+#include <stan/math/fwd/fun/beta.hpp>
+#include <stan/math/fwd/fun/digamma.hpp>
 #include <stan/math/fwd/core.hpp>
```
Comment (mcol) [inline stan/math/fwd/fun/inc_beta.hpp]: Ordering.

File: stan/math/fwd/fun/inverse.hpp [hunk 11]
```diff
@@ -1,12 +1,12 @@
...
 
 #include <stan/math/prim/err.hpp>
 #include <stan/math/prim/mat/fun/Eigen.hpp>
 #include <stan/math/prim/mat/fun/multiply.hpp>
 #include <stan/math/fwd/core.hpp>
-#include <stan/math/fwd/mat/fun/to_fvar.hpp>
-#include <stan/math/fwd/mat/fun/multiply.hpp>
+#include <stan/math/fwd/fun/to_fvar.hpp>
+#include <stan/math/fwd/fun/multiply.hpp>
 #include <stan/math/prim/mat/fun/inverse.hpp>
```
Comment (mcol) [inline stan/math/fwd/fun/inverse.hpp]: Ordering.

File: stan/math/fwd/fun/sum.hpp [hunk 12]
```diff
@@ -1,13 +1,37 @@
...
+/**
+ * Return the sum of the entries of the specified standard
+ * vector.
+ *
+ * @tparam T Type of vector entries.
+ * @param m Vector.
+ * @return Sum of vector entries.
+ */
+template <typename T>
+inline fvar<T> sum(const std::vector<fvar<T> >& m) {
```
Comment (mcol) [inline stan/math/fwd/fun/sum.hpp]: [optional] Change `> >` to `>>`.

File: test/unit/math/fwd/fun/value_of_rec_test.cpp [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+#include <stan/math/fwd/core.hpp>
+#include <gtest/gtest.h>
+#include <vector>
+
+TEST(AgradFwd, value_of_rec) {
+  using stan::math::fvar;
+
+  fvar<double> a = 5.0;
+  fvar<fvar<double> > ff_a(5.0);
+  fvar<fvar<fvar<fvar<double> > > > ffff_a(5.0);
```
Comment (mcol) [inline test/unit/math/fwd/fun/value_of_rec_test.cpp]: [optional] Change `> >` to `>>` in all this file.

File: test/unit/math/fwd/fun/value_of_test.cpp (hunk 7 above)
Comment (rok-cesnovar) [reply test/unit/math/fwd/fun/value_of_test.cpp]: Lets tackle this one separately. I do think it improves the code style. But a quick grep shows roughly a hundred files with the same style. This PR is large enough I think.

File: stan/math/fwd.hpp (hunk 8 above)
Comment (rok-cesnovar) [reply stan/math/fwd.hpp]: I used the same order as is used in the existing /arr.hpp, /mat.hpp and scal.hpp. See https://github.com/stan-dev/math/blob/develop/stan/math/fwd/arr.hpp, https://github.com/stan-dev/math/blob/develop/stan/math/fwd/scal.hpp, https://github.com/stan-dev/math/blob/develop/stan/math/fwd/mat.hpp

I however dont know the exact reasons behind the order. If you feel that this should be addressed we can discuss this here or on discourse.

File: stan/math/fwd/fun/sum.hpp (hunk 12 above)
Comment (rok-cesnovar) [reply stan/math/fwd/fun/sum.hpp]: See above.

File: test/unit/math/fwd/fun/value_of_rec_test.cpp (hunk 13 above)
Comment (rok-cesnovar) [reply test/unit/math/fwd/fun/value_of_rec_test.cpp]: See above.

File: stan/math/fwd.hpp (hunk 8 above)
Comment (mcol) [reply stan/math/fwd.hpp]: No big deal, it looked funny, that's all.

File: test/unit/math/fwd/fun/value_of_test.cpp (hunk 7 above)
Comment (mcol) [reply test/unit/math/fwd/fun/value_of_test.cpp]: Fair enough!

**Comment by rok-cesnovar** [comment]
Thanks for the fast turnaround @mcol !

**Pull request merged**